Support affine permutations in transforms

grlee77 opened the pull request:
# Description

closes #3410

This PR attempts to support cases where the affine is set using a permutation as in #3410. Prior to this PR, this could cause a linear algebra error as seen in that issue. Fixing the behavior is done here by adding some special handling code for permutations in the `set_slice` method for the Affine transform class and making corresponding updates to Layer methods as needed. From the perspective of the viewer, all attributes of `viewer.dims` reflect the axis order **after** the permutation and most things just continue to work as before. However, there are a few places, such as slicing of the original data array where we need to convert from the permuted order back to the original data order. So far these changes have been fairly small and mostly confined to the base Layer class (no changes have been needed in any of the VisPy or Qt-based classes).

As far as I can tell, things are working as expected in interactive use with the GUI, including the ray intersections for 3D views and the "values" reported in the status below the viewer window (i.e. `get_value` and `get_ray_intersections`). 

Aside from the permutation handling, I also added an `is_diagonal` property that allows optimizing some operations of the `Affine` class for the common case when the `linear_matrix` portion is just a diagonal matrix. That part could also be in a separate smaller PR if preferred.

There is also a bug fix related to 0.5 pixel offset in the ray intersection calculations. I could make a separate PR isolating just that change as well.

### alternative proposal

An alternative to the solution here is to manually transpose the data axes so that the affine becomes diagonal rather than remaining a permutation. This could be in, e.g. `add_image`. That would simplify things in terms of avoiding potential confusion between order of dimensions pre vs. post permutation as data and viewer dimensions would never be permuted relative to one another. There are a couple of drawbacks with that approach, though

1.) it is impractical to transpose the full dataset in cases involving large data, especially data that is larger than memory or stored in the cloud.
2.) by transposing and then updating the affine upon load, the user will likely not realize this was done. if the user then saves the data again, but uses the original affine, the transform information would be incorrect.

## Type of change
<!-- Please delete options that are not relevant. -->
- [ ] Bug-fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References
<!-- What resources, documentation, and guides were used in the creation of this PR? -->
<!-- If this is a bug-fix or otherwise resolves an issue, reference it here with "closes #(issue)" -->

# How has this been tested?
<!-- Please describe the tests that you ran to verify your changes. -->
- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

There are quite a few new tests covering the various additions to the `napari.utils.transform` classes. For testing with the ViewerModel and Layers, I have been doing this manually so far, verifying that things like the sliders cover the expected ranges and things work as expected. I need to add some unit tests for this, though, so have marked this as WIP. I also need to test a bit more with the other layer types aside from Image and Labels.

## Final checklist:
- [x My PR is the minimum possible work for the desired functionality
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

grlee77 committed:
add is_diagonal property

Affine implements a fast code path for transforms when the Affine is diagonal

grlee77 committed:
add is_diagonal docstring

grlee77 committed:
add is_permutation utility and test cases

grlee77 committed:
add is_permutation method to Transform classes

grlee77 committed:
update fast path in Affine.shear

grlee77 committed:
add perm property to Affine, CompositeAffine and expand tests for permutations

grlee77 committed:
support permutation in set_slice

grlee77 committed:
add expand_dims tests for Affine permutation matrix

grlee77 committed:
Handle permutations in layers

grlee77 committed:
docstring update

grlee77 committed:
fix expected output in test_world_extent_mixed_flipped

grlee77 committed:
Fix coordinate offset in Layer.get_ray_intersections

Fix handling of permuted dimensions in Layer.get_value

grlee77 committed:
minor change in _on_matrix_change code style

grlee77 committed:
fix: make sure indices and thumb_indices are tuples

jni commented on napari/layers/base/base.py:
```diff
@@ -634,6 +634,53 @@ def _dims_displayed_order(self):
+        if self._data_to_world.is_permutation:
+            perm = self._data_to_world.perm
+            return [perm[d] for d in self._dims_order]
+        return self._dims_order
```
I know some linters remove `else`s in situations like this, but I find it easier to read. In my first reading I'd missed the `return` inside the `if` block and it seemed like you were ignoring the work in that block.

```suggestion
        if self._data_to_world.is_permutation:
            perm = self._data_to_world.perm
            return [perm[d] for d in self._dims_order]
        else:
            return self._dims_order
```

Alternatively:

```python
        if self._data_to_world.is_permutation:
            perm = self._data_to_world.perm
            dims_order = [perm[d] for d in self._dims_order]
        else:
            dims_order = self._dims_order
        return dims_order
```

jni commented on napari/layers/base/base.py:
```diff
@@ -1401,6 +1471,9 @@ def get_ray_intersections(
         bounding_box = self._display_bounding_box(dims_displayed)
+        if not world:
+            # need 0 to be the corner of a pixel rather than its center
+            position = tuple(p + 0.5 for p in position)
```
Yeah, this should be a separate PR and issue, please. 0 *should* be the center of a pixel according to our coordinate conventions? (ie pixels should be considered samples taken at the given coordinates)

jni commented on napari/utils/transforms/transforms.py:
```diff
@@ -597,6 +648,28 @@ def expand_dims(self, axes: Sequence[int]) -> 'Affine':
+        )
+
+    @cached_property
+    def perm(self):
```
Can we rename this "permutation"? I don't think it's worth saving the 7 characters at the cost of readability.

grlee77 replied:
Yes, done!

jni reviewed (commented):
@grlee77 I've only had a chance to have a quick look. I'd definitely prefer for the 0.5 offset bug to be properly explained and fixed in a separate PR. Otherwise this looks ok to me, minus `.perm` -> `.permutation`.

Having said that, I wonder whether this could just be done with functions rather than adding methods/properties to the affine transform. `if (perm := get_permutation(tf)):`, where `get_permutation` returns `None` if the transform is not a permutation, and a permutation otherwise?

andy-sweet commented on napari/utils/transforms/transform_utils.py:
```diff
@@ -466,3 +466,86 @@ def is_matrix_triangular(matrix):
+        r, c = np.nonzero(np.abs(matrix) >= tol)
+    # permutation matrix will only have 1 entry per column
+    c = tuple(c)
+    c_check = len(list(set(c))) == n
```
```suggestion
    c_check = len(set(c)) == n
```

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -405,17 +421,33 @@ def ndim(self) -> int:
-        )
-        self._linear_matrix = compose_linear_matrix(rotate, scale, shear)
+        if self.is_diagonal:
+            for i in range(len(scale)):
```
You might want to do `scale = scale_to_vector(scale, ndim=self.ndim)` to ensure the scale is padded as in `compose_linear_matrix`.

grlee77 replied:
Good catch, done in [8825912](https://github.com/napari/napari/pull/4303/commits/8825912f1701cb63d62e60ba9b803bd940c34888)

andy-sweet commented on napari/components/_tests/test_layers_list.py:
```diff
@@ -486,18 +486,17 @@ def test_world_extent_mixed_ndim():
     layers.append(layer)
-    np.testing.assert_allclose(layer._data_to_world.scale, (1, -1))
-    np.testing.assert_allclose(layers.extent.step, (1, 1))
+    # scale and step reflect the values AFTER the affine has been applied.
```
I'm struggling to understand what this means exactly.

In the docstring of `Affine` we describe the transform as being the composition of a few components in a predefined order: `rotate * shear * scale + translate`.

Are we changing this order in general to something like `scale * rotate * shear * permutation + translate`?

grlee77 replied:
I think we still need to decide what we want this to look like. Right now there is a `data2physical` `CompositeAffine` transform which is `rotate * shear * scale + translate` like you say. Then there is also a separate `physical2world` transform that is a general `Affine` (although in practice only a diagonal physical2world works correctly prior to this PR). So far, I was testing with setting that separate `physical2world` matrix to be a permutation rather than the default of an identity matrix.

I think it would make sense that `data2physical` itself could involve a permutation, but haven't tried to refactor `CompositeAffine` to add a separate permutation property there.

grlee77 replied:
> but haven't tried to refactor CompositeAffine to add a separate permutation property there.

To clarify, there is the read-only permutation property as introduced in this PR that returns the axis order, but there is no permutation argument to the constructor. To assign a different permutation to the Affine after it has already been created, one would have to either assign `linear_matrix` or one could assign both `rotation` and `scale` as determined via `decompose_linear_matrix`

andy-sweet replied:
Thanks for the comments. I think I understand things a little better now, though that may have generated more clarifying questions elsewhere in this PR.

andy-sweet commented on napari/components/_tests/test_layers_list.py:
```diff
@@ -486,18 +486,17 @@ def test_world_extent_mixed_ndim():
+        np.random.random((15, 15)), affine=[[0, 1, 0], [2, 0, 0], [0, 0, 1]]
     )
     layers.append(layer)
-    np.testing.assert_allclose(layer._data_to_world.scale, (1, -1))
```
I'm curious how pure reflections are handled - i.e. what happens for `affine = [[1, 0, 0], [0, -1, 0], [0, 0, 1]]`? Does the reflection still get absorbed into `scale`? Can we have an explicit test for that if it doesn't already exist?

grlee77 replied:
yes, scale in this case would be [1, -1, 1]. I can add a test case with negative scales.

```Python
import numpy as np
from napari.utils.transforms import Affine

linear_matrix = np.asarray([[1, 0, 0], [0, -1, 0], [0, 0, 1]])
A = Affine(linear_matrix=linear_matrix)
print(A.scale)
```

This ends up being the same as the scale given by:
```Python
from napari.utils.transforms.transform_utils import decompose_linear_matrix

rotate, scale, translate = decompose_linear_matrix(linear_matrix)
```

andy-sweet replied:
A short extra test seems useful to me.

grlee77 replied:
added this test in [306c14a](https://github.com/napari/napari/pull/4303/commits/306c14a6aaa9268eadda07fdd7a563bf8e5e84a7)

andy-sweet reviewed (commented):
Thanks for your effort here @grlee77!

The tests are really helpful in understanding some of the changes here, and also help clarify the changes I don't understand yet.

I haven't reviewed all the changes yet because this is still draft and because I have some high level questions that are probably best answered before I continue.

In general, any PR that touches some of the code paths here (i.e. how transforms/layers/slicing interact) always confuses me a little, and unfortunately this one isn't an exception. That could definitely be down to my own ignorance and something I could fix. But I'm curious to know your perspective on how easy you found these changes to make and how happy you are with the solution. If everyone find PRs related to this logic hard to read and write, it suggests ensuring that redesign this logic is prioritized.

grlee77 committed:
move offset code into _ImageBase

grlee77 committed:
Merge branch 'fix-3d-offset-update' into add-transform-permutations

grlee77 committed:
Apply suggestions from code review

Co-authored-by: Juan Nunez-Iglesias <jni@fastmail.com>
Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

grlee77 committed:
apply scale_to_vector in all Affine.scale setter code paths

grlee77 committed:
Merge branch 'add-transform-permutations' of github.com:grlee77/napari into add-transform-permutations

grlee77 committed:
test a permutation with a negative entry, reflecting a flip along an axis

grlee77 committed:
rename perm -> permutation

grlee77 committed:
apply updates to is_diagonal from gh-4370

grlee77 committed:
make permutation attributes on Transform base class

change names to private:
_is_permutation and _permutation

grlee77 committed:
Merge remote-tracking branch 'upstream/main' into add-transform-permutations

grlee77 committed:
fix layers to use the private names

grlee77 committed:
add test cases for Image layer extent after permutation

grlee77 committed:
test viewer.dims attributes when dimensions are permuted

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
+    def _is_permutation(self) -> bool:
+        """Indicate a transform corresponding to a permutation of dimensions.
+
+        Can be overriden in subclasses to enable performance optimizations
```
```suggestion
        Can be overridden in subclasses to enable performance optimizations
```

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -135,11 +157,15 @@ def inverse(self) -> 'TransformChain':
+    def _is_diagonal(self) -> bool:
         if all(getattr(tf, '_is_diagonal', False) for tf in self):
             return True
         return getattr(self.simplified, '_is_diagonal', False)
```
```suggestion
        return all(tf._is_diagonal for tf in self)
```

As `_is_diagonal` is a property defined on the base class that has a default implementation, can we simplify this? Or maybe `return self.simplified._is_diagonal`?

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -135,11 +157,15 @@ def inverse(self) -> 'TransformChain':
 
+    @property
+    def _is_permutation(self) -> bool:
+        return getattr(self.simplified, '_is_permutation', False)
```
```suggestion
        return self.simplified._is_permutation
```

As `_is_permutation` is defined on the base class, can we simplify?

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
+    def _is_diagonal(self) -> bool:
         """Indicate when a transform does not mix or permute dimensions.
 
         Can be overriden in subclasses to enable performance optimizations
```
```suggestion
        Can be overridden in subclasses to enable performance optimizations
```

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -445,6 +485,12 @@ def scale(self, scale):
             for i in range(len(scale)):
                 self._linear_matrix[i, i] = scale[i]
+        elif self._is_permutation:
+            # assumes scale[i] is the scaling on axis i AFTER the transform
```
After what transform? After the permutation?

andy-sweet replied:
Does this effectively mean that when `_is_permutation` is true then we can always have `linear_matrix = scale * permutation`?

I.e. the `rotate` and `shear` components will always be identity?

andy-sweet commented on napari/utils/transforms/transform_utils.py:
```diff
@@ -490,3 +490,66 @@ def is_diagonal(matrix, tol=1e-8):
+    # permutation matrix will only have 1 entry per column
+    c = tuple(c)
+    c_check = len(set(c)) == n
+    if not c_check:
```
```suggestion
    if len(set(c)) != n:
```

Optional: simplify or at least consider a more meaningful name than `c_check`.

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
+
+    @property
+    def _permutation(self) -> Optional[Tuple[int]]:
+        """Indicate when a transform does not mix or permute dimensions.
```
The summary line seems to be copy/pasted from `_is_permutation`? Consider just removing the summary and keeping the rest.

andy-sweet commented on napari/layers/base/base.py:
```diff
@@ -754,7 +804,11 @@ def _slice_indices(self):
+            self._data_dims_not_displayed
+        )
+
+        # _dims_point is already permuted so don't use _permuted_dims_not_displayed
```
`_permuted_dims_not_displayed` does not exist, but `_data_dims_not_displayed` does. Either update the comment. Or maybe better consider renaming the new properties on `Layer` to start with `_permuted` instead of `_data`?

grlee77 replied:
conclusion during the last architecture working group meeting was not to do this renaming

andy-sweet commented on napari/utils/transforms/transforms.py:
```diff
@@ -96,16 +98,36 @@ def expand_dims(self, axes: Sequence[int]) -> 'Transform':
 
+    @property
+    def _is_permutation(self) -> bool:
+        """Indicate a transform corresponding to a permutation of dimensions.
```
I need a more clear/explicit description here. I think this returns true if the transforms is a "pure" permutation (that also allows scaling afterwards), but that does not allow things like rotation and shear that mix dimensions - is that right?

andy-sweet commented on napari/layers/image/image.py:
```diff
@@ -729,11 +771,11 @@ def _set_view_slice(self):
         else:
             self._transforms['tile2data'].scale = np.ones(self.ndim)
-            image_indices = self._slice_indices
+            image_indices = self._permute_indices(self._slice_indices)
```
If I understand correctly, `_slice_indices` has already absorbed any permutation, so is `_permute_indices` inverting that so we can index using the non-permuted dimensions?

andy-sweet replied:
Does this only need to happen for images? Have you tested with other layer types (e.g. points)?

grlee77 replied:
I think there is a bug here in the multiscale case (I'm pretty sure it should be `image = self.data[level][image_indices]` rather than `image = self.data[level][indices]`). I will add test coverage for permutations when the data is multiscale.

andy-sweet reviewed (changes requested):
Thanks for splitting up this PR and getting the simpler parts in first. I understand the changes here better now, but am still a little confused/scared by this PR.

grlee77 committed:
add test of world extent for pure reflection

grlee77 committed:
Apply suggestions from code review

Co-authored-by: Andy Sweet <andrew.d.sweet@gmail.com>

grlee77 committed:
fix _is_diagonal for TransformChain

The composite may be diagonal even if the individual ones are not

grlee77 committed:
fix reflection test case

grlee77 committed:
Merge remote-tracking branch 'upstream/main' into add-transform-permutations

grlee77 committed:
parametrize multiscale extent tests

add parameterization over use of permutations and a color channel

grlee77 committed:
Move shape permutation from extent to level_shapes

This will ensure level_shapes and downsampling_factors are also in
the viewer dimension space.

When setting corner_pixels, need to then update to index level_shapes
via the axis indices in the data space (not world space)

grlee77 committed:
use data dimensions when indexing point data

This is needed for support of permutations in Points layers

Fix Points.extent

fix drag-and-drop point selection

grlee77 committed:
Support permutations in Surface layers

grlee77 committed:
Support permutations for Tracks layers

grlee77 committed:
Support permutations for Vectors layers

grlee77 committed:
refactor _extent_data for Points

can permute the smaller mins, maxs arrays instead
